content/gocourse-2019-05-melbourne.slide

```diff
@@ -816,49 +816,56 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 .caption J. D. Frazer "Illiad", User Friendly - [[https://en.wikipedia.org/wiki/User_Friendly#Crud_Puppy][Crud Puppy]]
 
 
-* Errors (1/3)
+* Errors in Go
 
-"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
-
-The `error` type is a built-in interface:
+- `error` is a builtin interface ([[https://golang.org/src/builtin/builtin.go?s=11246:11286#L251][source: error]])
+- Check errors explicitly
+- The [[https://golang.org/pkg/errors/][`errors`]] package has functions for creating new error values
+- Using `errors.New()`:
 
-	type error interface {
-	    Error() string
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
+	}
+- Alternatively using `fmt.Errorf`:
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
 	}
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+* Errors in Go (continued)
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
+- Implement custom error types ([[https://golang.org/pkg/net/#AddrError][source: AddrError]]):
+	type AddrError struct {
+		Err  string
+		Addr string
+	}
+	func (e *AddrError) Error() string {
+		if e == nil {
+			return "<nil>"
+		}
+		s := e.Err
+		if e.Addr != "" {
+			s = "address " + e.Addr + ": " + s
 		}
-		return catStore.cats[id], nil
+		return s
+	}
+- Error values can be logged and printed:
+	if _, err := Sqrt(-1); err != nil {
+		log.Println(err)
 	}
 
+* Error Handling
 
-* Errors (2/3)
-
-"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
-
-Panic should be used when something went unexpectedly wrong.
-For instance, initialise a regular expression with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if the pattern is a string constant.
-`MustCompile` will `panic` at initialisation.
-
-	var re = regexp.MustCompile(`^[a-c]+$`)
-	var re2 = regexp.MustCompile(`^[a-c+$`) // panic
-
-
-Use `panic` rarely and judiciously in your own code!
-
-
-* Errors (3/3)
-
-- Error [[https://goplay.space/#h4c7gQxzQsN][create]], [[https://goplay.space/#mze1De36XjV][handle]], [[https://goplay.space/#h1kA3qKaO6B][custom error]]
-- [[https://goplay.space/#9vu6iJHXMHC][Panic]]
-- [[https://goplay.space/#pMbrba6y9cY][Defer]] with [[https://goplay.space/#pc68j1BPMxn][extended sample]]
+- Check for the presence of an error by [[https://goplay.space/#40Ho_HWiYrb][comparing with nil]]
+- [[https://goplay.space/#GUSay51SeQh][Comparing error values]]
```

Comment:
I don't feel strongly about this, but I'm still a bit confused: when would you ever compare errors in this way?

Reply:
Hmm Lol yes I did understand your comment. But my worry is what if people try to do this. Shouldn't we say we aren't allowed to do such things?

I can remove it. But I was worried people might compare local error values.

Ok I will remove that and instead just explain on the slide how error values are represented in memory.

Reply:
Done